compiler/system.c

```diff
@@ -53,23 +53,51 @@
 #include "proto.h"
 
 #ifdef __ELKS__
-void *fmemrealloc(void *q, size_t size)
+
+#define MAX_NEAR_ALLOC  64U   /* max size to allocate from near heap */
```

Comment:
Why is this so small? That's going to cause lots of far heap allocations with small data. I like your reduction to 512U in the mem.c file as a reasonable alternative to 2K, but I'm worried that 64 bytes is going to reintroduce the kernel out of heap errors we saw before when we have a bunch of tiny far allocations.

Reply:
We need this value parametrized in the mem api. This value was obtained experimentally, decreasing from 1024, the first that worked.

Reply:
> We need this value parametrized in the mem api.

Agreed. I was thinking of replacing MAX_NEAR_ALLOC with `unsigned int max_near_alloc = 512` in mem.c, adding the line `extern unsigned int max_near_alloc;` in mem.h, and then adding `max_near_alloc = 64;` in the tool init or main code, something like that. That would allow the same copy of mem.c to be used, but parameterized to a different value at startup.

It doesn't really matter exactly how we do it now, other than my comments are to try to reduce you or I changing it again when the C library code takes over. (Yes, I should have done this a lot earlier in the ELKS C library, but didn't realize just how tight things are and what you've had to go through to get things running!)

All that said, I am quite surprised at having to lower this value to 64 in order for C86 to work! This means that C86 must be allocating lots and lots of allocations > 64, which I find surprising. IIRC only the symbols buffer uses malloc directly in C86, the code generation and other large buffers use a BLKLEN allocation in order to reduce explicit malloc calls  (see my other comment about that).

I will add the capability to see what sizes malloc and fmemalloc are being called with so we can get to the bottom of what is going on with the many (different?) allocations the tools are doing. I've been traveling and haven't had as much time, but its coming!

Reply:
Btw, is there a hard 4096 bytes of maximum memory allocation in fmemalloc?

Reply:
> is there a hard 4096 bytes of maximum memory allocation in fmemalloc?

No, fmemalloc can allocate as much contiguous memory as may be found in the running system. That's why I find it so strange 4K can't be allocated right after C86 start. The only way this should fail is when the system no longer has 4K contiguous bytes available. With just /bin/init, sh, make and c86 running, I would think there would he a lot more memory available. In order to tell, we would need to run `meminfo`from another login while C86 is running, to inspect memory. I am assuming your ELKS system has 640k main memory?

Reply:
> > is there a hard 4096 bytes of maximum memory allocation in fmemalloc?
> 
> No, fmemalloc can allocate as much contiguous memory as may be found in the running system. That's why I find it so strange 4K can't be allocated right after C86 start. The only way this should fail is when the system no longer has 4K contiguous bytes available. With just /bin/init, sh, make and c86 running, I would think there would he a lot more memory available. In order to tell, we would need to run `meminfo`from another login while C86 is running, to inspect memory. I am assuming your ELKS system has 640k main memory?

Yes, 640k main memory.